### PR TITLE
superseded by #741: test(delta-harness): add harness-core framework and scenarios

### PR DESCRIPTION
**Pull request by mkuchenbecker**
## Superseded

This fork-backed PR is superseded by the same-repository [PR #741](https://github.com/linkedin/openhouse/pull/741). Do not reopen this PR. The replacement head and base branches both belong to `linkedin/openhouse`.

### CODE REVIEW

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala [hunk 1]
```diff
@@ -0,0 +1,1176 @@
...
+    layouts
+      .filter(layout =>
+        layout.label.endsWith("/parquet") ||
+          layout.label.endsWith("/orc"))
+      .flatMap { layout =>
+        val preparations = List(
+          TablePreparation(
+            layout.label,
+            createAndSeed(layout, 3)
+              .sql("ddl")(table => s"ALTER TABLE $table ADD COLUMN cc int")(),
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala]: what ddl does is unclear here. ITs a function.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala [hunk 2]
```diff
@@ -0,0 +1,1176 @@
...
+  import Rows._
+
+  val ddlConsumerCases: List[Plan.Case] =
+    layouts
+      .filter(layout =>
+        layout.label.endsWith("/parquet") ||
+          layout.label.endsWith("/orc"))
+      .flatMap { layout =>
+        val preparations = List(
+          TablePreparation(
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala]: Each preparation needs a description.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala [hunk 3]
```diff
@@ -0,0 +1,1176 @@
...
+import scala.util.control.NonFatal
+
+trait DmlScenarios extends ScenarioKit {
+  import Rows._
+
+  val ddlConsumerCases: List[Plan.Case] =
+    layouts
+      .filter(layout =>
+        layout.label.endsWith("/parquet") ||
+          layout.label.endsWith("/orc"))
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala]: Why not just ahve a list of layouts?

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala [hunk 4]
```diff
@@ -0,0 +1,1176 @@
...
+            "ddlConsume:typeWiden."),
+          TablePreparation(
+            layout.label,
+            createAndSeed(layout, 3)
+              .sql("ddl")(table =>
+                s"ALTER TABLE $table WRITE ORDERED BY ${Core.long0.columnName}")(),
+            "ddlConsume:writeOrder."),
+          TablePreparation(
+            layout.label,
+            createAndSeed(layout, 3)
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala]: Create and seed should be separate calls. What is seeded should be visible or standard.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala [hunk 5]
```diff
@@ -0,0 +1,1176 @@
...
+import org.apache.iceberg.exceptions.BadRequestException
+import org.apache.iceberg.exceptions.ValidationException
+import com.linkedin.openhouse.javaclient.exception.WebClientResponseWithMessageException
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+import scala.annotation.tailrec
+import scala.reflect.{ClassTag, classTag}
+import scala.util.control.NonFatal
+
+trait DmlScenarios extends ScenarioKit {
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala]: ORC and Parquet are not DML scanarios. They are jsut scenarios.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+    val json = org.apache.iceberg.SchemaParser.toJson(schema)
+    println(s"fork.colDefault.api: column-default API present, serialized schema JSON = $json")
+
+    // (a) The default is serialized into the schema JSON.
+    assert(json.contains("initial-default"),
+      s"expected SchemaParser to serialize 'initial-default' into the schema JSON, got: $json")
+    // (b) toJson takes no format-version argument, so the key serializes the same regardless of format version.
+    // (c) The value round-trips through fromJson then toJson.
+    val reparsed = org.apache.iceberg.SchemaParser.fromJson(json)
+    val json2 = org.apache.iceberg.SchemaParser.toJson(reparsed)
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala]: Just import. Apply this across all of these deep calls.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala [hunk 7]
```diff
@@ -0,0 +1,1903 @@
...
+
+  /**
+   * The DELETE that selects a null string. It applies to a preparation that already holds a row
+   * whose string column is null, and it removes exactly that row.
+   */
+  val nullStringRowTestCases: List[DmlTestCase[CoreTable.type]] = List(
+    deleteByNullCondition)
+
+  /**
+   * DELETE WHERE datepartition = '2024-01-01-00' removes the rows in that partition value, keeps
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/DmlScenarios.scala]: datepartition is a smell. Instead of that, its a date column that is partitioned. Adding the name partition to the col sows confusion.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+import scala.util.control.NonFatal
+
+// The fork cases. OpenHouse compiles and runs against LinkedIn's fork of Apache Iceberg, the
+// com.linkedin.iceberg artifacts this module depends on, and a case here pins a behavior the
+// Iceberg library decides rather than one the catalog exposes: the column-default path, the write
+// distribution default for a partitioned write, the output-file replication key, the read split
+// size, and the compaction plan. These behaviors have no catalog SQL surface of their own, so a
+// case reaches them through the Iceberg API or a Spark configuration and asserts the result a
+// caller can observe.
+trait ForkScenarios extends ScenarioKit {
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala]: forkscenarios  are Iceberg tests. they should all run agains the fork, so its unclear why this is a file vs denormalized into other features.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala (hunk 6 above)
Comment (mkuchenbecker) [reply integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala]: unaddressed

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+    assert(json2.contains("initial-default"),
+      s"expected 'initial-default' to survive the fromJson/toJson round trip, got: $json2")
+    println("fork.colDefault.api: initial-default serialized with no format-version argument and round-trips")
+  }
+
+  /**
+   * Reflectively builds an optional int NestedField carrying the given initial default. Returns
+   * None when the builder API is absent, so a caller can assert that absence directly.
+   */
+  private def buildDefaultedIntField(id: Int, name: String, dflt: Int): Option[org.apache.iceberg.types.Types.NestedField] = {
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala]: move default values to its own PR.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+
+  /**
+   * A partitioned write defaults write.distribution-mode to NONE, so every input task writes every
+   * partition it holds and one append produces up to (input tasks times partitions) data files.
+   * Under an explicit HASH distribution the writer shuffles rows so one task owns each partition,
+   * clustering the append to roughly one file per partition. Appending the same multi-task
+   * DataFrame into a 4-partition table under each mode therefore yields at least as many files
+   * under the default as under HASH. The file format is the parameter.
+   */
+  private def forkPartitionDistDefault(fmt: String)(ctx: Ctx): Unit = {
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala]: this test doesn't seem useful. is it? a write distribution mode test makes sense but not a None test. To me this is purely spark and nothing specific to iceberg being tested. That may be fine depending on how we phrase it.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+  /**
+   * OutputFileFactory exposes FILE_REPLICATION_FACTOR as "file-replication-factor", and a factory
+   * built with a replication factor stamps that key into the property map of the output files it
+   * creates. Writes made through the table afterward still return the correct rows. It is not a
+   * settable table property; it is the key HDFS reads to set block replication on an output file
+   * when a replication factor is supplied to the factory, and the delete-file write path is the one
+   * path that supplies one. Reflection reaches the builder and getProperties because some Iceberg
+   * artifacts leave them out of the public compiled API.
+   */
+  private def forkFileReplicationFactor(ctx: Ctx): Unit = {
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala]: REplicaiton factor should be tested. its not a fork, but a feature.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+
+  /**
+   * spark.sql.iceberg.split-size decides how the read path combines data files into read tasks.
+   * Over several small files, a large split size combines them into fewer read tasks and a tiny
+   * split size splits them into more, visible through rdd.getNumPartitions, and both reads return
+   * the same rows. The planner shows the same effect directly: a split size above the whole table
+   * plans one task group, and a split size below one file plans one group per file. The file format
+   * is the parameter.
+   */
+  private def forkSplitSize(fmt: String)(ctx: Ctx): Unit = {
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ForkScenarios.scala]: again, these are iceberg features not a "fork"

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/HazardReaderWriterScenarios.scala [hunk 13]
```diff
@@ -0,0 +1,654 @@
...
+import org.apache.iceberg.exceptions.BadRequestException
+import org.apache.iceberg.exceptions.ValidationException
+import com.linkedin.openhouse.javaclient.exception.WebClientResponseWithMessageException
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+import scala.annotation.tailrec
+import scala.reflect.{ClassTag, classTag}
+import scala.util.control.NonFatal
+
+// The copy-on-write reader, writer and hazard families. The reader and writer cases pin the
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/HazardReaderWriterScenarios.scala]: CoW vs MoR is the test. "hazard" makes no sense.

**Review by mkuchenbecker — COMMENTED** [review]
I would reindex on capabilities vs Hazard / Fork / etc. More files is fine, ideal case is a new file + a delta to integrate it.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/NestedTypesScenarios.scala [hunk 14]
```diff
@@ -0,0 +1,519 @@
...
+          nestedMergeInsertCase(preparation),
+          nestedDeleteByNestedFieldCase(preparation),
+          nestedNullValuesCase(preparation))
+      }
+
+  // Type-edge coverage (TypesTable).
+
+  /** One unpartitioned scalar-type table per file format. */
+  val typesLayouts: List[Layout] =
+    List("parquet", "orc", "avro").map(format => Layout(s"types-unpartitioned/$format", table =>
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/NestedTypesScenarios.scala]: constructing this list literal is suspect vs getting a standard "AllFileTypes". this is a nit for now, but a standard list is perferred.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/Env.scala [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+          case cleanupFailure: Throwable =>
+            startupFailure.addSuppressed(cleanupFailure)
+        }
+        throw startupFailure
+    }
+  }
+}
+
+object Main {
+  def main(args: Array[String]): Unit = {
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/Env.scala]: Main seems to ahve nothing to do with env.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/NegativeDdlScenarios.scala [hunk 16]
```diff
@@ -0,0 +1,656 @@
...
+          s"CREATE TABLE $scratchTable ($columnDefinitions) " +
+            s"USING $dataSource PARTITIONED BY (no_such_column) " +
+            s"TBLPROPERTIES ('write.format.default'='${preparation.label}')"))
+
+      table.spark.sql(s"DROP TABLE IF EXISTS $scratchTable")
+      assert(exception.getMessage.contains("no_such_column"))
+    }
+
+  /** The rejected DML statements, on the preparedCoreFormats preparations. */
+  val negativeCases: List[Plan.Case] =
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/NegativeDdlScenarios.scala]: compositions seem like they should be at teh top or bottom (top is preferred IMO). We can organize by public and private.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioChangelog.scala [hunk 17]
```diff
@@ -0,0 +1,219 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioChangelog.scala]: This should fail when you attempt to do so across RTAS to a previous table. Make sure RTAS PR capturs that.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioColumnTag.scala [hunk 18]
```diff
@@ -0,0 +1,43 @@
+package harness
+
+/**
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioColumnTag.scala]: Stack on top unless tags need RTAS testing.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioCompactionPlanning.scala [hunk 19]
```diff
@@ -0,0 +1,137 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioCompactionPlanning.scala]: can be stacked on top

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioEncryption.scala [hunk 20]
```diff
@@ -0,0 +1,44 @@
+package harness
+
+import java.nio.file.{Files, Paths}
+
+/**
+ * Encryption: the OSS build writes table data in plaintext, because OpenHouse delegates table-data encryption to an
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioEncryption.scala]: can be stacked on top unless there is the chance RTAS silently drops encryption.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioEncryption.scala [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * Preparation axes: the standard seeded core table in Parquet, which is the format whose footer carries the marker.
+ *
+ * Case families: one family contributing 1 case.
+ */
+trait EncryptionScenarios extends ScenarioKit {
+
+  /** The plaintext data-file case, on the standard seeded Parquet table. */
+  lazy val encryptionCases: List[Plan.Case] =
+    List(dataFilePlaintextCase(preparedStandardTable("parquet")))
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioEncryption.scala]: orc should be the default test choice or both orc and parquet. Both is better. I would jsut standardize to do both.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioIncrementalRead.scala [hunk 22]
```diff
@@ -0,0 +1,93 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioIncrementalRead.scala]: can be stacked on top unless RTAS should fail or something as a negative test case

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioMaintenance.scala [hunk 23]
```diff
@@ -0,0 +1,169 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioMaintenance.scala]: can be stacked on top

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioMetadataTable.scala [hunk 24]
```diff
@@ -0,0 +1,96 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioMetadataTable.scala]: can be stacked on top

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioNamespace.scala [hunk 25]
```diff
@@ -0,0 +1,48 @@
+package harness
+
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioNamespace.scala]: can be stacked on top

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioPartitionTransform.scala [hunk 26]
```diff
@@ -0,0 +1,155 @@
+package harness
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.types.StructType
+
+/**
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioPartitionTransform.scala]: can be stacked on top

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioProcedure.scala [hunk 27]
```diff
@@ -0,0 +1,113 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioProcedure.scala]: just delete this

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioRename.scala [hunk 28]
```diff
@@ -0,0 +1,68 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioRename.scala]: oooh, I think RTAS and rename are interesting to test together if they are not in the RTAS pr.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioScanPlanning.scala [hunk 29]
```diff
@@ -0,0 +1,115 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioScanPlanning.scala]: can be stacked on top

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioSortOrder.scala [hunk 30]
```diff
@@ -0,0 +1,61 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioSortOrder.scala]: rtas should test that after an RTAS sort order can be changed or dropped if its not already

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioStreaming.scala [hunk 31]
```diff
@@ -0,0 +1,213 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioStreaming.scala]: can be stacked on top

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioTimeTravel.scala [hunk 32]
```diff
@@ -0,0 +1,98 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioTimeTravel.scala]: RTAS definately needs time travel testing as a negative test case.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioWriteDistribution.scala [hunk 33]
```diff
@@ -0,0 +1,161 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioWriteDistribution.scala]: can be stacked on top

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioWriterCompatibility.scala [hunk 34]
```diff
@@ -0,0 +1,50 @@
+package harness
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/ScenarioWriterCompatibility.scala]: can be stacked on top

File: integrations/spark/delta-harness/src/test/scala/harness/CaseCatalogTest.scala [hunk 35]
```diff
@@ -0,0 +1,136 @@
...
+  private val expectedCaseCount = 1177
+  private val expectedCatalogSha256 =
+    "a10676c9fe0af5169459a0c9ad74eb2d005b7c7c63e2b7c7d4364bb7d8cc5bb9"
+
+  /**
+   * Every capability the standard catalog is built from, in the order Plan integrates them. This list is written out
+   * here rather than derived from `Plan.contributions`, so adding, dropping, renaming or reordering a capability fails
+   * this test until the intended catalog shape is restated.
+   */
+  private val expectedContributionNames = List(
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/test/scala/harness/CaseCatalogTest.scala]: this test is a tautaulogy.

File: integrations/spark/delta-harness/src/test/scala/harness/CaseCatalogTest.scala [hunk 36]
```diff
@@ -0,0 +1,136 @@
...
+import org.junit.jupiter.api.Assertions.{assertEquals, assertTrue}
+import org.junit.jupiter.api.Test
+
+/**
+ * Pins the ordered catalog: its size, its fingerprint, the uniqueness of its IDs, the capability naming rule, and the
+ * rule that every capability contributes exactly once. Reading the catalog does not execute a case or start Spark.
+ */
+final class CaseCatalogTest {
+  private val expectedCaseCount = 1177
+  private val expectedCatalogSha256 =
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/test/scala/harness/CaseCatalogTest.scala]: does this regenerate automatically?

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/scenarios/ScenarioDml.scala [hunk 37]
```diff
@@ -0,0 +1,1570 @@
...
+
+  /**
+   * The cases that address columns by name and never write a whole seed-shaped row, so they run on a preparation whose
+   * column list has grown beyond the seed rows.
+   */
+  lazy val testCasesCompatibleWithAnAddedColumn: List[DmlTestCase[CoreTable.type]] =
+    readTestCases ++ deleteTestCases ++ updateTestCases
+
+  /**
+   * Every DML case, with the partition-predicate DELETE marked as a known bug: the Spark and Iceberg rewrite crashes on
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/scenarios/ScenarioDml.scala]: "Marked as a known bug"

This is a stateful comment. Either there is a test that is disabled with a reason why (i.e. currently broken) or the test enabled. 

So if there is a crash, there should be a disabled test that crashes.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/scenarios/ScenarioDml.scala [hunk 38]
```diff
@@ -0,0 +1,1570 @@
...
+              s"WHERE ${Core.long0.columnName} >= 2 ORDER BY ${Core.long0.columnName}")
+          .collect()
+          .toSeq
+          .map(_.get(Core.long0))
+        val after = table.state
+
+        assert(
+          selected == before.rows.map(_.get(Core.long0)).filter(_ >= 2).sorted,
+          s"filtered read returned $selected")
+        assert(after == before, "a read leaves the rows and the snapshot count unchanged")
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/scenarios/ScenarioDml.scala]: I don't think asserting no mutation is useful.

File: integrations/spark/delta-harness/src/main/scala/harness/openhouse/scenarios/ScenarioDml.scala [hunk 39]
```diff
@@ -0,0 +1,1570 @@
...
+          .sql(
+            s"SELECT ${Core.long0.columnName} FROM ${table.name} " +
+              s"WHERE ${Core.long0.columnName} >= 2 ORDER BY ${Core.long0.columnName}")
+          .collect()
+          .toSeq
+          .map(_.get(Core.long0))
+        val after = table.state
+
+        assert(
+          selected == before.rows.map(_.get(Core.long0)).filter(_ >= 2).sorted,
```
Comment (mkuchenbecker) [inline integrations/spark/delta-harness/src/main/scala/harness/openhouse/scenarios/ScenarioDml.scala]: this assumes there is data in the table. This assumption should be called out.
